File: OrderBookAnalytics/CPlusPlusOrderBook/order_book.cpp
// Order book in C++

#include <iostream>
#include <map>
#include <unordered_map>
#include <queue>
#include <vector>
#include <memory>
#include <chrono>
#include <iomanip>
#include <algorithm>
#include <random>

// Using declarations to avoid std:: prefix
using namespace std;
using namespace std::chrono;

// Type aliases for cleaner code
using Price = int; // Price in cents
using Quantity = int; // Number of shares
using OrderId = uint64_t; // Unique order identifier. uint64_t is a fixed-width integer type (u = unsigned (only 0 and positive number), int = integer, 64 = 64 bits)

// ===== Order Structure =====
// This represents a single order in the order book.
// Structure is public by default
struct Order {
    OrderId id;
    Price price;
    Quantity quantity;
    bool is_buy; // True if the order is a buy order, false if it is a sell order.

    // Constructor - initializes an order
    Order(OrderId id_, Price p, Quantity q, bool buy)
        : id(id_), price(p), quantity(q), is_buy(buy) {}
};


// ===== Price Level =====
// Each price level contains all orders at that price.
class PriceLevel {
    private:
        Price price_;
        Quantity total_quantity_;

        // deque is a double-ended queue - used for FIFO order matching
        // O(1) add/remove at both ends
        // So we create a deque that contains pointers to Order objects, where pointers are used to avoid copying the Order object
        // And we use shared_ptr instead of Order* to avoid manual memory management
        deque<shared_ptr<Order>> orders_;

    public:
        // Constructor - intializes a price level with a given price
        PriceLevel(Price p) : price_(p), total_quantity_(0) {}

        // Add order to this price level (back of queue)
        void add_order(shared_ptr<Order> order) {
            orders_.push_back(order);
            total_quantity_ += order->quantity;
        }

        // Remove order from this price level (for cancellations)
        bool remove_order(OrderId order_id) {
            // Find the order with matching ID
            // find_if is a library function that searches through a container
            // orders_.begin() is the beginning of the deque, and orders_.end() is the end of the deque
            // We then have a Lambda function:
            //  [order_id] is a capture clause, captures the order_id variable from outside
            //  (const shared_ptr<Order>& o) is the parameter, each order in the deque
            //  { return o->id == order_id;} is the body, return true if this is the order we're looking for
            // We then check if we found it, where if 'it' is the end of the deque, we didn't find it
            auto it = find_if(orders_.begin(), orders_.end(),
                [order_id](const shared_ptr<Order>& o) {
                    return o->id == order_id;
                });

            if (it != orders_.end()) {
                total_quantity_ -= (*it)->quantity; // We have to dereference the iterator to get the order object
                orders_.erase(it);
                return true;
            }
            return false;
        }


        // Execute quantity from front of queue (FIFO matching)
        int execute_quantity(int qty_to_execute) {
            int executed = 0;

            while (!orders_.empty() && qty_to_execute > 0) {
                // Using & to avoid copying the order object
                auto& front_order = orders_.front();

                if (front_order->quantity <= qty_to_execute) {
                    // Fully execute this order
                    executed += front_order->quantity;
                    qty_to_execute -= front_order->quantity;
                    total_quantity_ -= front_order->quantity;
                    orders_.pop_front(); // Remove fully executed order
                } else {
                    // Partially execute
                    front_order->quantity -= qty_to_execute;
                    total_quantity_ -= qty_to_execute;
                    executed += qty_to_execute;
                    qty_to_execute = 0;
                }
            }

            return executed;
        }


        // Getters
        Price get_price() const { return price_; }
        Quantity get_total_quantity() const { return total_quantity_; }
        size_t get_order_count() const { return orders_.size(); } // Use size_t for unsigned integer as order size could be bigger than int
        bool is_empty() const { return orders_.empty(); }

};


// ===== Order Book =====
class OrderBook {

    private:
        // map keeps prices sorted automatically
        // For bids: highest price first (greater<Price>)
        // For asks: lowest price first (default less<Price>)
        map<Price, unique_ptr<PriceLevel>, greater<Price>> bid_levels_;
        map<Price, unique_ptr<PriceLevel>> ask_levels_;

        // Hash map for O(1) order lookup by ID
        // Maps order ID to pair of (is_buy, price)
        unordered_map<OrderId, pair<bool, Price>> order_location_;

        // Statistics
        uint64_t total_messages_ = 0;
        uint64_t total_volume_traded_ = 0;

        // For order ID generation
        OrderId next_order_id_ = 1000;

    public:
        // Add a new order to the order book
        bool add_order(OrderId id, bool is_buy, Price price, Quantity quantity) {
            total_messages_++;

            // Check if order ID already exists
            if (order_location_.find(id) != order_location_.end()) { // If the order ID is found in the map
                return false; // Duplicate order ID
            }

            // Create the order
            // make_shared creates a shared_ptr to the new Order object
            // auto is used to automatically deduce the type of the variable
            auto order = make_shared<Order>(id, price, quantity, is_buy);

            if (is_buy) {
                /// Add to bid side
                // If price level doesn't exist, create it
                if (bid_levels_.find(price) == bid_levels_.end()) {
                    bid_levels_[price] = make_unique<PriceLevel>(price);
                }
                bid_levels_[price]->add_order(order);
            } else {
                // Add to ask side
                if (ask_levels_.find(price) == ask_levels_.end()) {
                    ask_levels_[price] = make_unique<PriceLevel>(price);
                }
                ask_levels_[price]->add_order(order);
            }

            // Track order location for fast cancellation
            order_location_[id] = {is_buy, price};
            return true;
        }

        
        // Cancel an existing order
        bool cancel_order(OrderId id) {
            total_messages_++;

            // Find order location
            auto it = order_location_.find(id);
            if (it == order_location_.end()) {
                return false; // Order not found
            }

            // Get the is_buy and price from the second element of the pair, which is a pair of bool and Price
            bool is_buy = it->second.first;
            Price price = it->second.second;

            // Remove from appropriate side
            if (is_buy) {
                // Check if the price level exists
                if (bid_levels_.find(price) == bid_levels_.end()) {
                    order_location_.erase(it); // Clean up the orphaned entry
                    return false;
                }
                if (bid_levels_[price]->remove_order(id)) {
                    // If level is now empty, remove it
                    if (bid_levels_[price]->is_empty()) {
                        bid_levels_.erase(price);
                    }
                    order_location_.erase(it);
                    return true;
                } 
            } else {
                // Check if the price level exists
                if (ask_levels_.find(price) == ask_levels_.end()) {
                    order_location_.erase(it); // Clean up the orphaned entry
                    return false;
                }
                if (ask_levels_[price]->remove_order(id)) {
                    // If level is now empty, remove it
                    if (ask_levels_[price]->is_empty()) {
                        ask_levels_.erase(price);
                    }
                    order_location_.erase(it);
                    return true;
                }
            }

            return false; // Should never reach here
        }


        // Execute a market order (aggressive order that crosses spread for simplicity)
        int execute_market_order(bool is_buy, Quantity quantity) {
            total_messages_++;
            int executed = 0;

            if (is_buy) {
                // Buy market order - match against best ask
                while (!ask_levels_.empty() && quantity > 0) {
                    // Get best ask (lowest price)
                    auto& [price, level] = *ask_levels_.begin();

                    int level_executed = level->execute_quantity(quantity);
                    quantity -= level_executed;
                    executed += level_executed;
                    total_volume_traded_ += level_executed;

                    // Remove if level is empty
                    if (level->is_empty()) {
                        ask_levels_.erase(ask_levels_.begin());
                    }
                }
            } else {
                // Sell market order - match against best bid
                while (!bid_levels_.empty() && quantity > 0) {
                    // Get best bid (highest price)
                    auto& [price, level] = *bid_levels_.begin();

                    int level_executed = level->execute_quantity(quantity);
                    quantity -= level_executed;
                    executed += level_executed;
                    total_volume_traded_ += level_executed;

                    // Remove if level is empty
                    if (level->is_empty()) {
                        bid_levels_.erase(bid_levels_.begin());
                    }
                }   
            }

            return executed;
        }


        // Get the best bid price and quantity
        pair<Price, Quantity> get_best_bid() const {
            if (bid_levels_.empty()) {
                return {0, 0}; // No bids
            }

            // We use const auto& to avoid copying the price and level
            const auto& [price, level] = *bid_levels_.begin();
            return {price, level->get_total_quantity()};
        }

        // Get the best ask price and quantity
        pair<Price, Quantity> get_best_ask() const {
            if (ask_levels_.empty()) {
                return {INT32_MAX, 0}; // No asks
            }

            // We use const auto& to avoid copying the price and level
            const auto& [price, level] = *ask_levels_.begin();
            return {price, level->get_total_quantity()};
        }

        // Calculate the spread
        double get_spread() const {
            // As the pair are just two int values, & is not needed as these values are copyable
            auto [bid_price, bid_qty] = get_best_bid();
            auto [ask_price, ask_qty] = get_best_ask();

            if (bid_qty == 0 || ask_qty == 0) {
                return 0.0; // No spread
            }

            return (ask_price - bid_price) / 100.0; // Convert to dollars
        }

        // Calculate mid price
        double get_mid_price() const {
            auto [bid_price, bid_qty] = get_best_bid();
            auto [ask_price, ask_qty] = get_best_ask();

            if (bid_qty == 0 || ask_qty == 0) {
                return 0.0; // No mid price
            }

            return (bid_price + ask_price) / 200.0; // Convert to dollars
        }


        // Display order book state
        void display(int levels = 5) const {
            cout << "\nOrder book!" << endl;
            cout << "Messages processed: " << total_messages_ << endl;
            cout << "Total volume traded: " << total_volume_traded_ << " shares" << endl;

            // Display asks (in reverse order)
            vector<pair<Price, Quantity>> asks_to_show;
            int count = 0;
            for (const auto& [price, level] : ask_levels_) {
                if (count++ >= levels) break;
                asks_to_show.push_back({price, level->get_total_quantity()});
            }

            // Print asks from highest to lowest
            for (auto it = asks_to_show.rbegin(); it != asks_to_show.rend(); ++it) {
                cout << "Ask: $" << fixed << setprecision(2) << it->first / 100.0 << " x " << it->second << endl;
            }

            // Display bids
            count = 0;
            for (const auto& [price, level] : bid_levels_) {
                if (count++ >= levels) break;
                cout << "Bid: $" << fixed << setprecision(2) << price / 100.0 << " x " << level->get_total_quantity() << endl;
            }

            // Print spread
            cout << "Spread: $" << fixed << setprecision(2) << get_spread() << endl;

            cout << "===============\n" << endl;
        }

        // Get statistics
        void print_stats() const {
            cout << "\nPerformance stats!" << endl;
            cout << "Messages processed: " << total_messages_ << endl;
            cout << "Active orders: " << order_location_.size() << endl;
            cout << "Bid levels: " << bid_levels_.size() << endl;
            cout << "Ask levels: " << ask_levels_.size() << endl;
            cout << "Volume traded: " << total_volume_traded_ << " shares" << endl;

            auto [bid_price, bid_qty] = get_best_bid();
            auto [ask_price, ask_qty] = get_best_ask();

            if (bid_qty > 0) {
                cout << "Best bid: $" << fixed << setprecision(2) << bid_price / 100.0 << " x " << bid_qty << endl;
            }

            if (ask_qty > 0) {
                cout << "Best ask: $" << fixed << setprecision(2) << ask_price / 100.0 << " x " << ask_qty << endl;
            }

            cout << "Spread: $" << fixed << setprecision(2) << get_spread() << endl;
            cout << "Mid price: $" << fixed << setprecision(2) << get_mid_price() << endl;
            cout << "===============\n" << endl;
        }

        // Generate next order ID
        OrderId generate_order_id() {
            return next_order_id_++;
        }

};


// ===== Message Types =====
// Simple message structure
struct Message {
    enum Type {ADD, CANCEL, EXECUTE};
    Type type;
    OrderId order_id;
    bool is_buy;
    Price price;
    Quantity quantity;
};


// ===== Simulator =====
class MarketSimulator {
    private:
        OrderBook& book_;
        mt19937 rng_; // Random number generator
        uniform_real_distribution<> price_dist_; // Price distribution
        uniform_int_distribution<int> quantity_dist_; // Quantity distribution
        uniform_real_distribution<> action_dist_; // Action distribution

        Price base_price_ = 10000; // $100.00
        vector<OrderId> active_orders_; // Track orders we can cancel

    
    public:
        MarketSimulator(OrderBook& book)
            : book_(book),
              rng_(chrono::steady_clock::now().time_since_epoch().count()),
              price_dist_(0.01, 0.20), // Price offset from base price
              quantity_dist_(100, 1000), // Quantity range
              action_dist_(0.0, 1.0) {}

        // Generate and process random messages
        void simulate(int num_messages) {
            cout << "Simulating " << num_messages << " messages..." << endl;

            // First, populate the book with initial orders
            for (int i = 0; i < 10; ++i) {
                // Add bids
                Price bid_price = base_price_ - (i + 1) * 10; // 10 cents per level
                OrderId bid_id = book_.generate_order_id();
                Quantity bid_qty = quantity_dist_(rng_);
                if (book_.add_order(bid_id, true, bid_price, bid_qty)) {
                    active_orders_.push_back(bid_id);
                }

                // Add asks
                Price ask_price = base_price_ + (i + 1) * 10;
                OrderId ask_id = book_.generate_order_id();
                Quantity ask_qty = quantity_dist_(rng_);
                if (book_.add_order(ask_id, false, ask_price, ask_qty)) {
                    active_orders_.push_back(ask_id);
                }
            }

            cout << "Initial messages processed!" << endl;

            // Generate random messages
            for (int i = 0; i < num_messages; ++i) {
                double action = action_dist_(rng_);

                if (action < 0.5) {
                    // Add new order (50% chance)
                    add_random_order();
                } else if (action < 0.7) {
                    // Execute market order (20% chance)
                    execute_random_market_order();
                } else if (action < 0.9 && !active_orders_.empty()) {
                    // Cancel random order (20% chance)
                    cancel_random_order();
                } else {
                    // Add new order (10% chance)
                    add_random_order();
                }

                // Periodically display the order book
                //if ((i + 1) % 100 == 0) {
                //    cout << "Processed " << i + 1 << " messages..." << endl;
                //}
            }
        }


    private:

        void add_random_order() {
            bool is_buy = action_dist_(rng_) < 0.5;

            // Price around the base price
            Price price;
            if (is_buy) {
                price = base_price_ - static_cast<Price>(price_dist_(rng_) * 100); // 1-20% below base
                price = max(price, 1); // Ensure positive price
            } else {
                price = base_price_ + static_cast<Price>(price_dist_(rng_) * 100); // 1-20% above base
            }

            Quantity quantity = quantity_dist_(rng_);
            OrderId id = book_.generate_order_id();

            bool success = book_.add_order(id, is_buy, price, quantity);
            if (success) {
                active_orders_.push_back(id);
            }
        }

        void execute_random_market_order() {
            bool is_buy = action_dist_(rng_) < 0.5;
            Quantity quantity = quantity_dist_(rng_) / 2; // Smaller market orders

            int executed = book_.execute_market_order(is_buy, quantity);

            // Update base price based on executed quantity
            if (executed > 0) {
                if (is_buy) {
                    base_price_ += 1; // Price goes up on buying
                } else {
                    base_price_ -= 1; // Price goes down on selling
                }
            }
        }

        void cancel_random_order() {
            if (active_orders_.empty()) return;

            // Pick random order from active orders
            size_t size = active_orders_.size();
            if (size == 0) {
                return; // Double check
            }
            
            uniform_int_distribution<size_t> idx_dist(0, size - 1);
            size_t idx = idx_dist(rng_);
            
            // Bounds check
            if (idx >= active_orders_.size()) return;
            
            OrderId id = active_orders_[idx];

            bool cancel_success = book_.cancel_order(id);
            
            if (cancel_success) {
                // Remove from active orders - use safer method
                active_orders_.erase(active_orders_.begin() + idx);
            } else {
                // Order wasn't in book, but might still be in our list - remove it anyway
                active_orders_.erase(active_orders_.begin() + idx);
            }
        }
};


// ===== Performance Test =====
void performance_test(int num_orders) {
    cout << "\nPerformance test!" << endl;

    OrderBook book;

    // Measure time for adding orders
    auto start = high_resolution_clock::now();

    for (int i = 0; i < num_orders; ++i) {
        book.add_order(i, i % 2 == 0, 10000 + (i % 100), 100 + (i % 500));
    }

    auto end = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(end - start);

    cout << "Added " << num_orders << " orders in " << duration.count() << " microseconds" << endl;
    cout << "Average time per order: " << static_cast<double>(duration.count()) / num_orders << " microseconds per order" << endl;
    cout << "Throughput: " << (static_cast<double>(num_orders) * 1000000 / duration.count()) << " orders per second" << endl;

    // Test market orders
    start = high_resolution_clock::now();

    for (int i = 0; i < 100; ++i) {
        book.execute_market_order(i % 2 == 0, 50);
    }

    end = high_resolution_clock::now();
    duration = duration_cast<microseconds>(end - start);

    cout << "Executed 100 market orders in " << duration.count() << " microseconds" << endl;

    book.print_stats();
}


// ===== Main Function =====
int main() {
    cout << "C++ Order Book Simulation" << endl;

    // Create order book
    OrderBook book;

    // Create and run simulator
    cout << "Creating simulator..." << endl;
    MarketSimulator simulator(book);
    cout << "Starting simulation..." << endl;
    simulator.simulate(1000000);
    cout << "Simulation complete!" << endl;

    // Display final state
    cout << "Displaying book..." << endl;
    book.display();
    cout << "Printing stats..." << endl;
    book.print_stats();
    cout << "Done!" << endl;

    // Run performance test
    performance_test(10000000);

    return 0;
}

File: OrderBookAnalytics/ITCHOrderBook/order_book.py
""" This file contains the order book implementation. """

# Import used libraries
import time
import numpy as np
from collections import defaultdict
from typing import List, Dict, Optional, Tuple
import heapq
from data_strucs import Order, PriceLevel


class ITCHOrderBook:
    """ ITCH Order Book with individual order tracking. """

    def __init__(self, symbol: str = "TEST"):

        self.symbol = symbol

        # Order storage - flat dictionary for O(1) lookup
        self.orders: Dict[int, Order] = {}

        # Price levels - sorted dictionaries
        # Using integer prices (cents) for performance
        self.bid_levels: Dict[int, PriceLevel] = {}
        self.ask_levels: Dict[int, PriceLevel] = {}

        # Sorted price tracking for best bid/ask
        # Using heaps for O(log n) best price operations
        self.bid_prices: List[int] = [] # Max heap (negative values)
        self.ask_prices: List[int] = [] # Min heap

        # Cache best prices for O(1) access
        self._best_bid: Optional[int] = None
        self._best_ask: Optional[int] = None

        # Performance tracking
        self.operation_times = defaultdict(list)
        self.message_count = 0

        # Track order book events
        self.last_update_time = time.time_ns()
        self.total_volume_traded = 0


    def add_order(self, order_id: int, timestamp: int, side: str, price: int, shares: int) -> bool:
        """
        Add a new order to the order book.
        """

        start_time = time.perf_counter_ns()

        # Check for duplicates
        if order_id in self.orders:
            return False
        
        # Create order
        order = Order(order_id, timestamp, side, price, shares)
        self.orders[order_id] = order

        # Add to appropriate side
        if side == 'BID':
            if price not in self.bid_levels:
                self.bid_levels[price] = PriceLevel(price, 'BID')
                heapq.heappush(self.bid_prices, -price) # Negative for max heap

            self.bid_levels[price].add_order(order_id, shares)

            # Update chached best bid
            if self._best_bid is None or price > self._best_bid:
                self._best_bid = price

        else: # ASK
            if price not in self.ask_levels:
                self.ask_levels[price] = PriceLevel(price, 'ASK')
                heapq.heappush(self.ask_prices, price) # Min heap

            self.ask_levels[price].add_order(order_id, shares)

            # Update cached best ask
            if self._best_ask is None or price < self._best_ask:
                self._best_ask = price

        # Update
        self.message_count += 1
        self.last_update_time = timestamp

        # Track performance
        elapsed = time.perf_counter_ns() - start_time
        self.operation_times['add_order'].append(elapsed)

        return True
    

    def execute_order(self, order_id: int, executed_shares: int) -> bool:
        """
        Execute an partial or full order.
        """
        start_time = time.perf_counter_ns()

        if order_id not in self.orders:
            return False
        
        order = self.orders[order_id]
        order.shares -= executed_shares
        self.total_volume_traded += executed_shares

        # Update price level volume
        if order.side == 'BID':
            if order.price in self.bid_levels:
                self.bid_levels[order.price].total_volume -= executed_shares
        else: # ASK
            if order.price in self.ask_levels:
                self.ask_levels[order.price].total_volume -= executed_shares

        # Remove if fully executed
        if order.shares <= 0:
            self.delete_order(order_id)

        # Update
        self.message_count += 1

        # Track performance
        elapsed = time.perf_counter_ns() - start_time
        self.operation_times['execute_order'].append(elapsed)

        return True
    

    def cancel_order(self, order_id: int, cancelled_shares: int) -> bool:
        """
        Partial cancel of an order.
        """

        start_time = time.perf_counter_ns()

        if order_id not in self.orders:
            return False
        
        order = self.orders[order_id]
        order.shares -= cancelled_shares

        # Update level volume
        if order.side == 'BID':
            if order.price in self.bid_levels:
                self.bid_levels[order.price].total_volume -= cancelled_shares
        else: # ASK
            if order.price in self.ask_levels:
                self.ask_levels[order.price].total_volume -= cancelled_shares

        # Remove if fully cancelled
        if order.shares <= 0:
            self.delete_order(order_id)

        # Update
        self.message_count += 1

        # Track performance
        elapsed = time.perf_counter_ns() - start_time
        self.operation_times['cancel_order'].append(elapsed)

        return True
    

    def delete_order(self, order_id: int) -> bool:
        """
        Delete an order from the order book.
        """

        start_time = time.perf_counter_ns()

        if order_id not in self.orders:
            return False
        
        order = self.orders[order_id]

        # Remove from price level
        if order.side == 'BID':
            if order.price in self.bid_levels:
                level = self.bid_levels[order.price]
                if level.remove_order(order_id, order.shares):
                    # Remove level if empty
                    if level.order_count == 0:
                        del self.bid_levels[order.price]
                        # Update best bid if necessary
                        if order.price == self._best_bid:
                            self._update_best_bid()
        else: # ASK
            if order.price in self.ask_levels:
                level = self.ask_levels[order.price]
                if level.remove_order(order_id, order.shares):
                    # Remove level if empty
                    if level.order_count == 0:
                        del self.ask_levels[order.price]
                        # Update best ask if necessary
                        if order.price == self._best_ask:
                            self._update_best_ask()

        # Remove order
        del self.orders[order_id]
        self.message_count += 1

        # Track performance
        elapsed = time.perf_counter_ns() - start_time
        self.operation_times['delete_order'].append(elapsed)

        return True
    

    def get_best_bid(self) -> Optional[Tuple[int, int]]:
        """ Get best bid price and total volume at that level. """
        if self._best_bid and self._best_bid in self.bid_levels:
            level = self.bid_levels[self._best_bid]
            return (self._best_bid / 100.0, level.total_volume)
        
        return None
    

    def get_best_ask(self) -> Optional[Tuple[int, int]]:
        """ Get best ask price and total volume at that level. """
        if self._best_ask and self._best_ask in self.ask_levels:
            level = self.ask_levels[self._best_ask]
            return (self._best_ask / 100.0, level.total_volume)
        
        return None
    
    
    def get_spread(self) -> Optional[float]:
        """ Get bid-ask spread in dollars. """
        if self._best_bid and self._best_ask:
            return (self._best_ask - self._best_bid) / 100.0
        
        return None
    

    def get_mid_price(self) -> Optional[float]:
        """ Get mid price (average of best bid and ask) in dollars. """
        if self._best_bid and self._best_ask:
            return (self._best_bid + self._best_ask) / 200.0
        
        return None
    

    def get_queue_position(self, order_id: int) -> Optional[int]:
        """
        Get queue position for specific order.
        """

        if order_id not in self.orders:
            return None
        
        order = self.orders[order_id]

        if order.side == 'BID':
            if order.price in self.bid_levels:
                return self.bid_levels[order.price].get_queue_position(order_id)
        else:
            if order.price in self.ask_levels:
                return self.ask_levels[order.price].get_queue_position(order_id)
        
        return None
    

    def get_book_depth(self, levels: int = 5) -> Dict:
        """
        Get order book depth up to N levels.
        """

        # Get top N bid prices
        top_bids = sorted(self.bid_levels.keys(), reverse=True)[:levels]
        bid_depth = [(price/100.0, self.bid_levels[price].total_volume, self.bid_levels[price].order_count) for price in top_bids]

        # Get top N ask prices
        top_asks = sorted(self.ask_levels.keys())[:levels]
        ask_depth = [(price/100.0, self.ask_levels[price].total_volume, self.ask_levels[price].order_count) for price in top_asks]

        return {
            'bids': bid_depth,
            'asks': ask_depth,
            'bid_levels': len(self.bid_levels),
            'ask_levels': len(self.ask_levels),
            'total_orders': len(self.orders),
        }
    

    def _update_best_bid(self) -> None:
        """ Update cached best bid. """
        if self.bid_levels:
            self._best_bid = max(self.bid_levels.keys())
        else:
            self._best_bid = None
    

    def _update_best_ask(self) -> None:
        """ Update cached best ask. """
        if self.ask_levels:
            self._best_ask = min(self.ask_levels.keys())
        else:
            self._best_ask = None


    def get_performance_stats(self) -> Dict:
        """ Get performance statistics. """

        stats = {
            'total_messages': self.message_count,
            'active_orders': len(self.orders),
            'bid_levels': len(self.bid_levels),
            'ask_levels': len(self.ask_levels),
            'total_volume_traded': self.total_volume_traded
        }

        # Add timing statistics
        for operation, times in self.operation_times.items():
            if times:
                times_ns = np.array(times)
                stats[f'{operation}_avg_ns'] = np.mean(times_ns)
                stats[f'{operation}_p50_ns'] = np.percentile(times_ns, 50)
                stats[f'{operation}_p99_ns'] = np.percentile(times_ns, 99)

        return stats